Fix #7579: Tree: provide PF Selenium - wrapper and integrationtests

FlipWarthog opened the pull request:
Took a first pass at Tree. @melloware and I bounced some ideas back and forth regarding the best way to access the various elements within the p:tree. It's currently rough, but any feedback would be greatly appreciated, then I can incorporate them and build out some more tests. @christophs78

FlipWarthog committed:
Add initial test classes

FlipWarthog committed:
Add base page

FlipWarthog committed:
Merge branch 'master' into PF3887-test

FlipWarthog committed:
Merge remote-tracking branch 'origin/master' into PF3887-test

FlipWarthog committed:
Merge remote-tracking branch 'origin/master' into PF3887-test

FlipWarthog committed:
Merge remote-tracking branch 'origin/master' into PF3887-test

FlipWarthog committed:
Merge branch 'primefaces:master' into PF3887-test

FlipWarthog committed:
Merge branch 'primefaces:master' into PF3887-test

FlipWarthog committed:
Implement base test

FlipWarthog committed:
Add TreeNode selenium and initial click

FlipWarthog committed:
Merge remote-tracking branch 'origin/master' into PF3887-test

FlipWarthog committed:
Merge remote-tracking branch 'origin/master' into PF3887-test

FlipWarthog committed:
Add actual first test for Toggle

melloware commented on primefaces-selenium/primefaces-selenium-components/src/main/java/org/primefaces/selenium/component/model/tree/TreeNode.java:
```diff
@@ -0,0 +1,73 @@
+    private String childSelector;
+
+    public TreeNode(WebElement webElement, String selector) {
+        this.webElement = webElement;
```
You might want to pass the owning Tree component into this as well.  You will probably need it.

FlipWarthog replied:
Added, as well as the parent `TreeNode`, if available.

tandraschko commented:
feels like RootNode should be removed (selector moved to Tree component) and reuse TreeNode instead

melloware commented on primefaces-selenium/primefaces-selenium-components/src/main/java/org/primefaces/selenium/component/model/tree/TreeNode.java:
```diff
@@ -0,0 +1,73 @@
+    public void setSelector(String selector) {
+        this.selector = selector;
+    }
+
```
What about a `toggle()` convenience method right here that calls `getTreeToggler.click()`.   The reason this will be good is for another test you will need to do this with AJAX "expand" and "collapse" behaviors and we will need to guard this `toggle`

FlipWarthog replied:
Added `toggle()` and `select()`, both with `guardAjax(...)`.

melloware commented on primefaces-selenium/primefaces-selenium-components/src/main/java/org/primefaces/selenium/component/model/tree/RootNode.java:
```diff
@@ -0,0 +1,54 @@
+    public static final String CHILD_SELECTOR = ".ui-tree-container>li";
+    private WebElement webElement;
+
+    public RootNode(WebElement webElement) {
```
Pass the owning Tree component in here maybe?

christophs78 commented:
Regarding (integration-)testing:
I personally try to add some AJAX-postbacks (via eg CommandButton).
With this postbacks we ensure correct re-rendinger of componentes (is still the same value / node selected?). And i also like to add some (info-)message(s) with the selected value / node from the ViewBean. And validate this messages as part of the integration-test.
Similar things for ajax-events.

melloware commented:
Yep we will definitely need tests that use the "expand" and "collapse" AJAX events with messages etc.

FlipWarthog committed:
Merge remote-tracking branch 'origin/master' into PF3887-test

FlipWarthog committed:
Refacor based on PR review\
\
Add AJAX based tests for toggle/select\
Remove unneeded RootNode class\
Add convenience methods to TreeNode

FlipWarthog commented:
Comments addressed. Let me know if there's anything else. @melloware @tandraschko @christophs78

christophs78 commented on primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/tree/Tree002Test.java:
```diff
@@ -0,0 +1,94 @@
+
+        // Act Pt. 2
+        Actions actions = new Actions(page.getWebDriver());
+        actions.keyDown(Keys.META).click(first.getLabel()).keyUp(Keys.META).perform();
```
Maybe we should add some PrimeSelenium.guardAjax to avoid running into timing-issues.

FlipWarthog replied:
I can wrap it around buttonShowSelectedNode, but if I try something like `actions.keyDown(Keys.META).click(PrimeSelenium.guardAjax(first.getLabel())).keyUp(Keys.META).perform();` Selenium seems to lock up when it performs the series of actions.

christophs78 commented:
Thanks for working on this!

The one thing i´m missing is something like
`<p:commandButton id="buttonUpdate" value="Update" update="@form"/>`
And after this some re-evaluation of the Tree.
Why? We should ensure state is kept through (AJAX-)update-cycles. (Sound´s trivial but is something new components sometimes fail because no one cares. And should not break after some refactoring or other changes.)

FlipWarthog committed:
Add guardAjax around button clicks

FlipWarthog committed:
Tree: add integration test for Ajax update

FlipWarthog commented:
@christophs78 Added a basic test for the AJAX update. There's still a ton of Tree functionality to test, specifically multi-select with both `List<TreeNode>` and `TreeNode[]` fields in the backing bean (which is what kicked off this exercise). I think I'll save those for a separate issue, if that's okay.

christophs78 commented on primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/tree/Tree001Test.java:
```diff
@@ -78,13 +80,64 @@ public void testBasicAndToggling(Page page) {
+        Assertions.assertEquals("Movies", third.getLabelText());
+
+        // Act
+        PrimeSelenium.guardAjax(page.buttonUpdate).click();
```
Please expand (toggle) one node before clicking update-button. So we can check this node is still expanded after update.

FlipWarthog replied:
Added.

FlipWarthog committed:
Expand node before AJAX update to ensure it is still expanded

FlipWarthog committed:
Merge remote-tracking branch 'origin/master' into PF3887-test

melloware commented:
I just kicked off the GitHub Action tests to see if this runs and passes. If So I will merge. @christophs78 we can then modify and build other tests in other PR's if that is OK with you?

melloware commented:
Looks like we got 1 failure.  Most likely need an AJAX guard somewhere
```
 Errors: 
Error:    Tree002Test.testSingleSelection:73->AbstractTreeTest.assertMessage:41 » StaleElementReference
[INFO] 
Error:  Tests run: 360, Failures: 0, Errors: 1, Skipped: 1
```

christophs78 commented:
> I just kicked off the GitHub Action tests to see if this runs and passes. If So I will merge. @christophs78 we can then modify and build other tests in other PR's if that is OK with you?

Yes it´s fine. We can always add more. :-)

FlipWarthog committed:
Fix stale element reference in selection test

FlipWarthog commented:
@melloware I added a commit to address the stale element reference, if you want to kick the build again.
